### DISCUSSION TITLE
How to control the ah1s in unreal engine

### DISCUSSION THREAD

**Discussion by Tommy-T1984** [JSBSim-Team/jsbsim#659] [Support requests]
I was set the aircraft model ah1s,but it cant control it. It need set some new keymapping?

**Comment by bcoconni** [comment]
This topic is neither a bug report nor a feature request, it will be addressed in the Discussion tab.
Issue #656 is now closed.

Regarding your question, you'll need to hack the UE plugin to interface it with an helicopter (see issue #625 for a similar issue).

**Comment by gallonmate** [comment]
I personally plan to add support for helicopters in the UE plugin. But it will be a few weeks-months before I get around to it, including multiplayer, etc. Mostly likely someone will add support soon.

**Comment by hongzhou91** [comment]
In UE5,I set JSBSimMovement Aircraft model is ah1s and play the level.I try to control the helicopter  take off and land, but press the keyboard 4, the throttle is full, it is still on the ground, press the forward keyboard 8, the nose is tilted forward on the ground and spins in place. I want to know how it controls the helicopter take-off and landing and normal flight forward and backward.

**Comment by seanmcleod** [comment]
What keyboard input have you connected to the helicopter's collective?

**Reply by hongzhou91** [reply]
Hi，I use key mappings from https://github.com/JSBSim-Team/jsbsim/tree/master/UnrealEngine

**Comment by seanmcleod** [comment]
And what is the key mapping to the collective control of the helicopter?

**Reply by hongzhou91** [reply]
I dont konw
I guess press ctrl+W Toggle Engines Mixture On then press3 or 4 contrl Throttle.
Helicopter doesn't respond.

**Comment by seanmcleod** [comment]
If you're planning on flying a helicopter you need to understand what the **collective** control is. Then in terms of the UE-JSBSim plug-in you need to figure out what the keyboard mapping if any there is between the keyboard and the **collective** control.

https://github.com/JSBSim-Team/jsbsim/blob/5fa0a2f534e5b3f41384abd2f98b50590b2a535d/aircraft/ah1s/Systems/rotor_control.xml#L68-L72

**Reply by MariusMyburg** [reply]
Good day. I am also trying to get the AH-1H to work. But I seem to be missing some fundamental understanding of JSBSim, because I do not understand why I am getting the behavior I am getting.

1. I created a Pawn, added a Static Mesh and the JSBSim component.
2. I specified the ah1h Aircraft Model.
3. I position the plane in my level based on the debug hints that are drawn, just a meter or so above the ground.
4. I click play, expecting the pawn to fall to the ground. I specified no inputs whatsoever. Debug info shows no collective and engines off. Instead of falling to the ground, my pawn shoots up into the air very fast.

I have seen something very similar with fixed wing aircraft I tried. Engines off, yet they would move backwards on begin play.

Could you please help me understand what I am doing wrong?

Thank you.

**Reply by MariusMyburg** [reply]
PS: when this happened, I set 'Start With Engines Running' to off, hoping it would fix it, but it does not.

**Reply by hongzhou91** [reply]
Hi，I also tried this way，the AH1S still not takeoff and land.

**Reply by hongzhou91** [reply]
I think maybe setting the throttle separately

**Reply by gallonmate** [reply]
Helicopter control is not yet supported by the UE plugin. It will be added sometime in the future.

**Reply by hongzhou91** [reply]
If I want to add this feature, what can I do? Have some suggestion?

**Comment by seanmcleod** [comment]
> Helicopter control is not yet supported by the UE plugin. It will be added sometime in the future.

Yep, but anyone can add support for it in the meantime.

> If I want to add this feature, what can I do? Have some suggestion?

Have you looked at my comments about understanding the helicopter controls in general and then making sure you have a mapping from keyboard to the relevant inputs in the helicopter model?

Also take a look through the AH1S files to familiarize yourself on how the model works in terms of the required inputs etc. I did post one relevant snippet above.

Also have you taken a look at the AH1S sample script that is included in the repo which includes script commands to start the helicopter and have it take-off etc.?

https://github.com/JSBSim-Team/jsbsim/blob/master/scripts/ah1s_flight_test.xml

**Reply by hongzhou91** [reply]
Sorry I m newer and not very familiar with C++，I read the documentation and I understand is it to set a pitch key in the blueprint to give the offset the ups and downs?

**Reply by seanmcleod** [reply]
https://en.wikipedia.org/wiki/Helicopter_flight_controls

**Comment by gallonmate** [comment]
So I hardcoded the commands for the collective control into both JSBSim and UE. The collective now works but I don't think this is the intended way for controlling new vehicles in JSBSim is it? The UE plugin is written with hardcoded controls and I'm not that experienced with c++ to figure out how to define/add new controls at runtime, since it looks like the helicopter flight model declares new controls, mainly the collective cmd.
Example snippet of UE plugin currently. Basically I'm not sure how to access new/all properties of FCS at runtime, without hardcoding like shown.
![image](https://user-images.githubusercontent.com/28702753/174526753-0ac504d6-8f0b-4107-9eea-736c80a282d2.png)

**Reply by hongzhou91** [reply]
I am not a programmer, has yours been successful? I want to know how to make it，It seem set  FCS->SetClmd(Commands.Collective) and save it,then add keyboard Axis Mappings?

**Reply by gallonmate** [reply]
No it only partially worked. More controls need to be coded in c++. So instead I'm adding a new Blueprint Node to access all commands/settings. This way all settings, commands, properties of JSBSim can be accessed without c++. I will post it on Git in a few days when it's done.
![image](https://user-images.githubusercontent.com/28702753/174995217-0f6a1d3d-b980-4399-8d8b-6cac88c3bd2c.png)

**Reply by seanmcleod** [reply]
What other controls were needed? When I glanced at the script I mentioned above - https://github.com/JSBSim-Team/jsbsim/blob/master/scripts/ah1s_flight_test.xml the only other one I noticed was the enabling of the RPM governor.

```xml
<set name="fcs/rpm-governor-active-norm" value="1.0" action="FG_RAMP" tc="5.0"/>
```

**Reply by gallonmate** [reply]
Yup that's most likely it! But I became focused on other aspects of the plugin, such as adding this node and updated the plugin to run jsb 120hz independent of the game's frame rate.

**Reply by hongzhou91** [reply]
Okay, I'll wait for your release,

**Reply by hongzhou91** [reply]
@seanmcleod Sir, Would you please tell me what other control elements needed to be added? And What value the Collective variable needed? The helicopter  moves only when I  set it to 1. But it move very strangely.

**Reply by hongzhou91** [reply]
@seanmcleod @gallonmate need to add collective-ctrl-rad?

**Reply by gallonmate** [reply]
@hongzhou91 I finished adding the new function to set any command. You can copy the plugin from my repo and also see the update readme with section for the new functions/commands. You can start by setting fcs/rpm-governor-active-norm to 0.1, then slowly increase collective-cmd-norm. Also this helicopter is hard to fly with keyboard :laughing: 
![multiplecommands](https://user-images.githubusercontent.com/28702753/175456775-ccd64b32-625b-4444-9c81-5aed20e0f660.PNG)

**Reply by hongzhou91** [reply]
> @hongzhou91 I finished adding the new function to set any command. You can copy the plugin from my repo and also see the update readme with section for the new functions/commands. You can start by setting fcs/rpm-governor-active-norm to 0.1, then slowly increase collective-cmd-norm. Also this helicopter is hard to fly with keyboard 😆 ![multiplecommands](https://user-images.githubusercontent.com/28702753/175456775-ccd64b32-625b-4444-9c81-5aed20e0f660.PNG)

Thanks! But I am not programmer. So I didn't find the above picture after copying it, did I add it manually?
make a JSBsim Movement Component Blueprint and then add functions command console batch?

**Reply by gallonmate** [reply]
I added three new blueprint nodes to the plugin, that was it. The picture is an example of how to use the new nodes, which I put in the readme. https://github.com/gallonmate/jsbsim/blob/master/UnrealEngine/README.md#extended-commands-and-properties

**Reply by hongzhou91** [reply]
I can't find the blueprint you descript in your fork branch, have you not upload completely yet?

**Reply by gallonmate** [reply]
There is no blueprint and I'm not adding a blueprint. I added 3 new blueprint functions shown in the first picture in Readme under the Extended Commands section I linked. The two other blueprint pictures are just example of how to use the new functions. If you need help using Unreal Engine, there are many tutorials or ask for help in an Unreal Engine community.Here is a picture of the only thing I added to the UE plugin, you can right click in a blueprint and search for them. ![image](https://user-images.githubusercontent.com/28702753/175575742-72e84be0-9ea1-4393-b170-1c8748e2b1a1.png)

**Reply by hongzhou91** [reply]
> No it only partially worked. More controls need to be coded in c++. So instead I'm adding a new Blueprint Node to access all commands/settings. This way all settings, commands, properties of JSBSim can be accessed without c++. I will post it on Git in a few days when it's done. ![image](https://user-images.githubusercontent.com/28702753/174995217-0f6a1d3d-b980-4399-8d8b-6cac88c3bd2c.png)

Why the JSBSimMovement componet object reference is not compatible with self object reference?
the command property node functions create input argument?

**Reply by Tommy-T1984** [reply]
how to add the feature by c++?

**Reply by gallonmate** [reply]
What? I updated the plugin to include a new blueprint node so you can access any controls you want. If you want the new blueprint nodes you have to re-download the UE plugin and recompile.

**Reply by hongzhou91** [reply]
> What? I updated the plugin to include a new blueprint node so you can access any controls you want. If you want the new blueprint nodes you have to re-download the UE plugin and recompile.

Are you making a new blueprint class called Actor Component? 
How did you make the JSBSim Movement Component and node get JSBSim
![1](https://user-images.githubusercontent.com/62862985/184324692-36cbad73-1ce0-42a7-bcfa-d5c117f9eff9.png)

**Reply by gallonmate** [reply]
@hongzhou91 No I didn't make any new blueprint class. I only used/tested inside the original airplane blueprint called BP_Airliner that is included in the plugin. You need to add the Component JSBSim, like adding static mesh component. Click Add in the top left corner of blueprint. But if you want to make a new blueprint from scratch you need to look at BP_Airliner as example and also read the tutorial/guide https://dev.epicgames.com/community/learning/tutorials/mmL/a-diy-flight-simulator-tutorial for any issues.

**Comment by seanmcleod** [comment]
> but I don't think this is the intended way for controlling new vehicles in JSBSim is it?

What do you mean? 

The other option as opposed to adding a new method to `FCS` is to use the Property Manager and set the property `fcs/collective-cmd-norm`.

**Comment by gallonmate** [comment]
I only meant I don't know which is the preferred or ideal method, ie should I do a pull request adding helicopter controls to jsbsim. But this Property Manager sounds interesting, I will have to see if I can find an example of use case that will fit UE plugin.

**Comment by seanmcleod** [comment]
Take a look at - https://github.com/JSBSim-Team/jsbsim/discussions/627#discussioncomment-2817930 for some example Property Manager code.

**Comment by gallonmate** [comment]
Thank you sir! And lol there is a lot of properties. I'll probably make a command/menu UI out of this.
![image](https://user-images.githubusercontent.com/28702753/174687456-bf22244d-e558-4d27-b5df-843807cadbe4.png)

**Comment by seanmcleod** [comment]
> And lol there is a lot of properties.

Sure, but only a handful that are pilot inputs. 

If you were writing a generic flight simulator then you would typically present some UI to allow the user to map keyboard, joystick axes, joystick buttons etc. to pilot inputs.

**Comment by hongzhou91** [comment]
I found the AH1S is so difficult to fly, I turn the governor to 0.1 and increase the collective, and when the RPM is up to 7000, the helicopter lifts up, and to left a bit. When the helicopter is in the sky, it is so difficult to keep it in balance

**Comment by seanmcleod** [comment]
> difficult to keep it in balance

Welcome to helicopter flying :wink:

If you want to make it simpler to fly then add a AFCS (Automated Flight Control System) to the model, e.g. you could implement one that maps pilot input on the yaw pedals to a yaw rate, so if there is no pilot input on the yaw pedals, e.g. the pilot only moves the collective then the AFCS will automatically output yaw pedal movements to keep the yaw rate = 0.

**Reply by hongzhou91** [reply]
The AFCS file is systems/afcs.xml? and how to install it.

> > difficult to keep it in balance
> 
> Welcome to helicopter flying 😉
> 
> If you want to make it simpler to fly then add a AFCS (Automated Flight Control System) to the model, e.g. you could implement one that maps pilot input on the yaw pedals to a yaw rate, so if there is no pilot input on the yaw pedals, e.g. the pilot only moves the collective then the AFCS will automatically output yaw pedal movements to keep the yaw rate = 0.

**Comment by seanmcleod** [comment]
I wasn't aware of https://github.com/JSBSim-Team/jsbsim/blob/master/systems/afcs.xml

But looking at https://github.com/JSBSim-Team/jsbsim/blob/master/aircraft/ah1s/ah1s.xml it looks like it's already included by the AH1S model.

So I suggest you read through - https://github.com/JSBSim-Team/jsbsim/blob/master/aircraft/ah1s/Systems/rotor_control.xml and `afcs.xml` to become familiar with what they provide and how you can tune them to see whether that makes the helicopter easy enough to fly. If not then you could try adding an AFCS mode as I suggested to see if that helps more.

Or as the comment in `rotor_control.xml` mentions :wink:

> There is also a (!really serious!) alternative for a 'fully augmented' helicopter - just
     use flightgears' UFO fdm and stop bothering about the stuff above ;)

**Reply by hongzhou91** [reply]
”Inputs from AFCS/SAS system       
ap/collective-cmd
ap/elevator-cmd
 ap/aileron-cmd
ap/rudder-cmd“
Is that them，it looks like they want to add input keys？

**Comment by bomber64** [comment]
I'm curious, could I download this AH1s for unreal and have a look at it ?

I'm thinking I could steer you in the right direction.

**Comment by seanmcleod** [comment]
> I'm curious, could I download this AH1s for unreal and have a look at it ?

Do you mean @hongzhou91's Unreal blueprint mapping user keyboard etc. inputs to the AH1S's control inputs? The AH1S model is the one that's included in the JSBSim repo.

**Reply by bomber64** [reply]
What I'm asking for is the ability to fly this helicopter for myself in unreal and then see what comes of it.

**Comment by bomber64** [comment]
"The AH1S model is the one that's included in the JSBSim repo."

Where would I find this then ?

**Reply by seanmcleod** [reply]
https://github.com/JSBSim-Team/jsbsim/tree/master/aircraft/ah1s

**Reply by bomber64** [reply]
Thanks for this.

I see the problem.

**Reply by seanmcleod** [reply]
What's the problem you see?

The initial problem was that the Unreal component didn't have a mapping for the collective in particular from Unreal inputs to the JSBSim model's pilot inputs. Once that was implemented then there was a comment that the helicopter was difficult to fly.

> I found the AH1S is so difficult to fly, I turn the governor to 0.1 and increase the collective, and when the RPM is up to 7000, the helicopter lifts up, and to left a bit. When the helicopter is in the sky, it is so difficult to keep it in balance

I suggested if they wanted easier control they should look at making use of an AFCS.

**Reply by bomber64** [reply]
Have you flown it in Flightgear ?
What's your observations ?

**Reply by seanmcleod** [reply]
I haven't flown the AH1S in FlightGear.

**Reply by bomber64** [reply]
I have and this flight model is basically unflyable.

It's not about the FCS but everything else.

I shall now put my tin hat on.

**Comment by seanmcleod** [comment]
I'm not aware of any pre-built JSBSim-Unreal binaries with the AH1S model included and working. So your best bet would be to clone the repo, follow the instructions to build the Unreal components, test that you can run and fly the default airliner model. Then look into switching to the AH1S model by either modifying the Unreal blueprint etc. or ask @hongzhou91 for the changes they made.

**Reply by bomber64** [reply]
If that's my best bet, which seems a bit  'long winded'. My involvement in helping unreal guys is going to be limited.

And unreal guys if you want a working helicopter flight model get in touch

**Comment by gallonmate** [comment]
@bomber64 
Do you mean a JSBSim helicopter flight model or separate helicopter system for Unreal. Either way why not just share it publicly?

As for everyone else that reads and comments here.
The AH1S is 100% flyable. As Sean said you need to use the AFCS (it's group of various settings for stabilizing/auto pilot system and not a single switch/setting). It especially should be used during take off. Pretty sure it says so right in the model's file. Though I will say the AH1S is not perfect helicopter model or something for beginners. But it is 100% flyable using the various AFCS settings.

As for why I never mentioned it is flyable: I'm not here to hand hold someone into making a helicopter video game using JSBSim. The Unreal plugin/project is a reference and a starting point. It's not really meant for beginners to programming or game development. 

Anyways, as for using the AFCS I couldn't find any real world data on how the AH1S's stability system works or how a pilot would generally make use of it. I mean the stability forces/values to be used, etc. I looked through the real world AH1S manual and it's a general overview of the stability controls. Obviously a pilot can do whatever he wants while in flight and completely turn it off. For example you can turn on all the AFCS related settings and take off vertically with just some pilot corrections. But then you need to disable or turn down the value of certain AFCS settings in order to full control again in flight. I didn't want to show someone how to fly the AH1S with AFCS only to be asked MORE questions. If people can't figure out certain things on their own, well, I'm not going to do everything for them. I apologize for my ranting.

So now that I had my quick rant, here are the generic controls enabling and disabling a few of the AFCS commands which will allow stable flight with the AH1S.
![image](https://user-images.githubusercontent.com/28702753/193757898-d7c75bb6-493d-4b9a-bcb0-b6e906e3a328.png)

As seen in the picture, I use a simple button/key press to enable and and disable the following AFCS commands:
ap/afcs/yaw-channel-active-norm
ap/afcs/pitch-channel-active-norm
ap/afcs/roll-channel-active-norm

There are other ap and afcs related commands and settings for even greater control/stability, if you just spend some time researching. Also these commands can have for example values of between 0 and 1. So you can have full strength stabilization at 1, at the expense of less pilot control. Or you can have a weak stabilization using 0.2 or 0.5 and allow for more pilot control. Generally speaking you keep them enabled at full strength for take off, landing, and general forward flight. And disable it to make extreme maneuvers.

I'll also now mention the AH1S was basically the first(second) helicopter created for JSBSim, developed by one guy over 10 years ago. He laid the ground work but not much has happened since and there's little to no documentation. If you read flightgear forums, JSBSim helicopters are a controversial topic.

There are a few newer JSBSim helicopter models for flightgear if you search, that could be edited to work with Unreal/JSBSim and would likely be easier to fly than the AH1S. There is a R22 even on github if you search.

**Reply by bomber64** [reply]
https://forum.outerra.com/index.php?topic=4976.0

**Reply by bomber64** [reply]
Same question to you gallonmate.

Have you flown the AH1 in Flightgear ?

**Reply by seanmcleod** [reply]
Are you making a specific distinction between Flightgear versus the JSBSim+Unreal combination? Given the work done by @gallonmate to get the AH1S model working with Unreal documented above I'm 99% certain it's been flown by @gallonmate using the JSBSim+Unreal combination at least :wink:

**Reply by bomber64** [reply]
If the only way to get it to fly in Unreal was to add additional FCS functionality either in JSBsim or within Unreal itself, then pointing to the flight model in JSBsim repository is only showing me half the story.

So I asked a simple question as to whether a person had flown it in Flightgear (in it's unmodified format)

**Reply by gallonmate** [reply]
Thanks for sharing the link. And no I haven't actually used Flightgear before. Edit. I'm using 100% unmodified JSBSim and unmodified AH1S. If you enable the same afcs settings, as I did in Unreal, you should be able to "fly" it. To be clear I am not claiming the AH1S is a good or bad flight model. Only that I was able to "fly" it without crashing by toggling these commands on:
ap/afcs/yaw-channel-active-norm
ap/afcs/pitch-channel-active-norm
ap/afcs/roll-channel-active-norm

Does Flightgear have a property tree editor/window where you can access and enable these commands? These same commands should be available in Flightgear once the AH1S is loaded/running. In JSBSim a value of 1.0 is 100% enabled for those commands. They need be on for most of the flight.

**Reply by seanmcleod** [reply]
But the original AH1S model in JSBSim includes AFCS functionality by default, see my comments I made above in this thread:

> I wasn't aware of https://github.com/JSBSim-Team/jsbsim/blob/master/systems/afcs.xml

> But looking at https://github.com/JSBSim-Team/jsbsim/blob/master/aircraft/ah1s/ah1s.xml it looks like it's already included by the AH1S model.

So no one is having to add additional FCS functionality. The first issue with the Unreal integration is that there was simply no mapping of some controls like collective at all between the Unreal inputs and the JSBSim control inputs. Then @gallonmate added additional toggle functionality etc. to toggle some of the AFCS options on and off etc. So in all this no new FCS functionality was added to the AH1S model that exists in the JSBSim repo.

**Reply by bomber64** [reply]
The problem with a helicopter is that the pilot is constantly making micro adjustments based on the motion of the helicopter which he's feeling in his gut. In a sim we don't have that feedback, so we're 'looking' to see what's occurring and then reacting. Unfortunately this puts us behind the curve and results in our actions adding to the problems. So it's not wrong to add code that simulated the pilot making these micro adjustments.

**Reply by seanmcleod** [reply]
Sure, but even real helicopters with pilots experiencing real-time sensations often include SAS (Stability Augmentation Systems) to make them easier and less tiring to fly. A SAS implementation will also make it easier to fly a fixed based simulator where the simulator pilot isn't receiving inner ear motion/acceleration cues. Either with the actual helicopter's SAS gains or more exaggerated SAS gains.

The AFCS system included in the JSBSim repo and used by the AH1S model is parametrized and can implement simple rate damping functionality to provide a SAS .

https://github.com/JSBSim-Team/jsbsim/blob/f13d015dfad49be306e4f313b5e7d1406205ea64/systems/afcs.xml#L7-L13

**Reply by bomber64** [reply]
Then my suggestion would be that the default setting of the SAS system within this FDM outputs the basic piloting instincts experienced by real-time sensations. 

Such that upon start-up the 'player' has some chance.

We're not out to trick anyone, and the feedback I got from a real life AH1 pilot about this FDM was not good.

**Comment by bomber64** [comment]
Another link

http://www.thejabberwocky.net/viewforum.php?f=15&sid=a79c6865d24d4702239b35f891877cce

**Reply by hongzhou91** [reply]
Hi bomber64, I've been busy with work, but I haven't succeeded
I also want to know the aerodynamic operation settings of this helicopter, which can be studied and discussed together
